File: solucoes/c/ex11_inverter_encadeada.c
/*
 * ex11_inverter_encadeada.c — solução
 *
 * Inverte uma ListaEncadeada (do repositório) religando os nós
 * já existentes, sem alocar nenhum nó novo.
 * Compile junto com lista_encadeada.c — veja o Makefile desta pasta.
 */
#include <stdio.h>
#include "lista_encadeada.h"

void inverter(ListaEncadeada *lista) {
    No *anterior = NULL;
    No *atual = lista->inicio;

    lista->fim = lista->inicio; /* o antigo inicio vira o novo fim */

    while (atual != NULL) {
        No *proximo = atual->proximo; /* guarda o resto da lista antes de sobrescrever */
        atual->proximo = anterior;    /* inverte a seta deste no */
        anterior = atual;             /* avanca 'anterior' */
        atual = proximo;              /* avanca 'atual' */
    }

    lista->inicio = anterior; /* o ultimo no visitado vira o novo inicio */
}

int main(void) {
    ListaEncadeada l;
    lista_criar(&l);

    lista_inserir_fim(&l, 10);
    lista_inserir_fim(&l, 20);
    lista_inserir_fim(&l, 30);

    printf("antes:    ");
    lista_percorrer(&l);

    inverter(&l);

    printf("\nesperado: 30, 20, 10\n");
    printf("depois:   ");
    lista_percorrer(&l);

    /* confirma que 'fim' tambem ficou consistente apos a inversao */
    int ultimo;
    lista_obter(&l, lista_tamanho(&l) - 1, &ultimo);
    printf("\nlista->fim aponta para o valor %d (deve ser 10)\n", ultimo);

    lista_destruir(&l);
    return 0;
}

File: aula01-estruturas-lineares/solucoes/python/ex11_inverter_encadeada.py
"""
ex11_inverter_encadeada.py — solução

Inverte uma ListaEncadeada (do repositório) religando os nós já
existentes, sem criar nenhum nó novo.
"""
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "python"))

from lista_encadeada import ListaEncadeada


def inverter(lista):
    anterior = None
    atual = lista.inicio

    lista.fim = lista.inicio  # o antigo inicio vira o novo fim

    while atual is not None:
        proximo = atual.proximo  # guarda o resto da lista antes de sobrescrever
        atual.proximo = anterior  # inverte a seta deste no
        anterior = atual  # avanca 'anterior'
        atual = proximo  # avanca 'atual'

    lista.inicio = anterior  # o ultimo no visitado vira o novo inicio


if __name__ == "__main__":
    l = ListaEncadeada()
    l.inserir_fim(10)
    l.inserir_fim(20)
    l.inserir_fim(30)

    print(f"antes:    {l}")
    inverter(l)
    print("esperado: ListaEncadeada([30, 20, 10])")
    print(f"depois:   {l}")

    print(f"\nl.fim aponta para o valor {l.fim.valor} (deve ser 10)")
